File: 백준/Silver/10164. 격자상의 경로/격자상의 경로.py
import sys
lambda input: sys.stdin.readLine().strip()

def factorial(n):
    if n == 0 or n == 1:
        return 1
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result

n, m, k = map(int, input().split())
sero = max(0, (k - 1)) // m
garo = max(0, (k - 1)) % m

way1 = factorial(sero + garo) // (factorial(sero) * factorial(garo))
way2 = factorial((n - 1 - sero) + (m - 1 - garo)) // (factorial(n - 1 - sero) * factorial(m - 1 - garo))
print(way1 * way2)

File: 백준/Silver/10164. 격자상의 경로/격자상의 경로.java
import java.io.*;
import java.util.*;

public class Main {

    static long factorial(int n) {
        long result = 1;
        for (int i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        int sero = Math.max(0, (k - 1)) / m, garo = Math.max(0, (k - 1)) % m;
        long way1 = factorial(sero + garo) / (factorial(sero) * factorial(garo));
        long way2 = factorial((n - 1 - sero) + (m - 1 - garo)) / (factorial(n - 1 - sero) * factorial(m - 1 - garo));
        System.out.print(way1 * way2);
    }
}